providers/config/providerConfig.go

```diff
@@ -8,24 +8,42 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
+	"os/exec"
 	"strings"
 
 	"github.com/DisposaBoy/JsonConfigReader"
 	"github.com/TomOnTime/utfutil"
 )
 
-// LoadProviderConfigs will open the specified file name, and parse its contents. It will replace environment variables it finds if any value matches $[A-Za-z_-0-9]+
+// LoadProviderConfigs will open or execute the specified file name, and parse its contents. It will replace environment variables it finds if any value matches $[A-Za-z_-0-9]+
 func LoadProviderConfigs(fname string) (map[string]map[string]string, error) {
 	var results = map[string]map[string]string{}
-	dat, err := utfutil.ReadFile(fname, utfutil.POSIX)
-	if err != nil {
-		// no creds file is ok. Bind requires nothing for example. Individual providers will error if things not found.
-		if os.IsNotExist(err) {
-			fmt.Printf("INFO: Config file %q does not exist. Skipping.\n", fname)
-			return results, nil
+
+	var dat []byte
+	var err error
+
+	filename := strings.TrimPrefix(fname, "!")
+	if stat, statErr := os.Stat(filename); statErr != nil {
+		return nil, statErr
+	} else if mode := stat.Mode(); mode&0111 == 0111 || strings.Contains(fname, "!") {
+		// found executable bit or was marked as executable by user
+		dat, err = executeCredsFile(filename)
```

Comment:
This logic seems very complex.  It works, but the next person to read it might appreciate more simple logic.

Suggestion: (1) Abstract out reading the creds file into readCredsFile() like you did with executeCredsFile(), (2) abstract out whether or not the creds file is executable to a function isExecutable(), (3) simplify logic such as:

```
    if strings.HasPrefix(fname, "!") {
        dat, err = executeCredsFile(strings.TrimPrefix(fname, "!"))
        if err != nil { ... }
   } else if isExecutable(fname) {
        dat, err = executeCredsFile(fname)
        if err != nil { ... }
   } else {
        dat, err = utfutil.ReadFile(fname, utfutil.POSIX)
        if err != nil { ... }
   }
```